src/lib/string_pool.rs

```diff
@@ -0,0 +1,506 @@
+            }
+
+            // Should always be safe and inbounds to offset into the buffer
+            unsafe {
+                cap_begin = start.add(len);
+                cap_end = start.add(cap) as *mut ICHAR;
+            }
+
+            // NOTE: We avoid UB when writing to the capacity via a mutable slice by zeroing the
+            // capacity ahead of time in the grow() method
+            let mut writeBuf = ExpatBufRefMut::new(cap_begin, cap_end);
+
+            let convert_res = unsafe {
+                XmlConvert!(
+                    enc,
+                    &mut readBuf,
+                    &mut writeBuf,
+                )
+            };
+
+            // TODO: How to not need wrapping_offset_from here? Cast to u/isize first?
+            let new_len = writeBuf.as_ptr().wrapping_offset_from(start).try_into().unwrap();
+
+            self.set_len(new_len);
+
+            if convert_res == XML_CONVERT_COMPLETED || convert_res == XML_CONVERT_INPUT_INCOMPLETE {
+                break;
+            }
+
+            // if !self.grow() {
+            //     return false;
+            // }
+
+            len = new_len;
+        }
+
+        true
+    }
+
+    pub(crate) unsafe fn copyString(
+        &self,
+        mut s: *const XML_Char,
+    ) -> Option<&mut [XML_Char]> {
+        // self.appendString(s);?
+        loop {
+            if !self.appendChar(*s) {
+                return None;
+            }
+            let fresh78 = s;
+            s = s.offset(1);
+            if *fresh78 == 0 {
+                break;
+            }
+        }
+
+        Some(self.consume_current_vec())
+    }
+
+    pub(crate) unsafe fn copyStringN(
+        &self,
+        mut s: *const XML_Char,
+        mut n: c_int,
+    ) -> Option<&[XML_Char]> {
+        // REVIEW: Is this correct:
+        // if self.ptr.is_null() && !self.grow() {
+        if self.is_empty() && !self.grow() {
+            /* The following line is unreachable given the current usage of
+            * poolCopyStringN().  Currently it is called from exactly one
+            * place to copy the text of a simple general entity.  By that
+            * point, the name of the entity is already stored in the pool, so
+            * pool->ptr cannot be NULL.
+            *
+            * If poolCopyStringN() is used elsewhere as it well might be,
+            * this line may well become executable again.  Regardless, this
+            * sort of check shouldn't be removed lightly, so we just exclude
+            * it from the coverage statistics.
+            */
+            return None;
+        }
+        while n > 0 {
+            if !self.appendChar(*s) {
+                return None;
+            }
+            n -= 1;
+            s = s.offset(1)
+        }
+
+        Some(&*self.consume_current_vec())
+    }
+
+    /// There's currently no try_push in Bumpalo, so can't determine if
+    /// it's possible to allocate or not
+    pub(crate) fn grow(&self) -> bool {
```

Comment:
There's still the call to it in `storeEntityValue` remaining

Reply:
Fair enough, I agree that we should keep that one for now since it could change behavior. In practice I doubt it will matter, but there's also not really a downside.